Design.md

```diff
@@ -309,3 +332,34 @@ Get the platform usage of global (without query parameter), tenant (with tenant_
         * `key` : `offset` and `value` : `marker`
     1. `AssumeRoleCache` for all the `User` and `S3 Credential` APIs
         * `key` : `RoleArn` and `value` : `Credentials`
+
+## SecretKey Encryption Strategy
+
+1. Secret keys are encrypted and decrypted using a cipher algorithm and secret password provided in the `crypto.yml` config file.
+1. A `key-id` property will be provided for each `secret-password` in the `crypto.yml` config file to maintain the version of the key and will be used in the Key rotation.
+1. A java object of class `SecretKeyRepoData` will be created with variables `keyId`, `encryptedData` and `cipherInformation`
+    * `keyId` variable will be of type String and stores the `key-id` property.
+    * `encryptedData` will be a byte array and stores the encrypted data bytes.
+    * `cipherInformation` will be an encapsulation to store all the additional information used by the cipher algorithm.
+        * Example: For `AES256GCM` algorithm, `cipherInformation` encapsulates the `nonce` variable, which will be a byte array and stores the nonce bytes.
+1. Java object of class `SecretKeyRepoData` will be serialized and stored as binary in Redis
+1. During decryption, `key-id` in the meta information can be used to identify the cipher algorithm used.
+1. Initially, only the AES256GCM cipher algorithm is supported. (The Secretbox algorithm is suggested for future releases)
+1. An ability to change the cipher algorithm will be provided.
+1. Example `crypto.yml`
+    ```yaml
+    keys:
+      - id: 1
+        cipher: AES256GCM
+        secretKey: dGhpc2lzYXJlYWxseWxvbmdhbmRzdHJvbmdrZXk=
+    ```
+1. **Key Rotation:**
+    * Key rotation is used to replace a key with another one once the original key is compromised or too old to use.
+    * `crypto.yml` file will be updated with a new key slot.
```

Comment:
Need to specify the 'order' of slots in the `keys` list. Should the 'new' key go first? Or last?

Reply:
Fixed in 1779625

> `crypto.yml` file will be updated with a new key slot on top of `keys` list with an incremented `id` value.